env/io_posix.cc

```diff
@@ -125,7 +126,45 @@ size_t GetLogicalBufferSize(int __attribute__((__unused__)) fd) {
 #endif
   return kDefaultPageSize;
 }
-} //  namespace
+
+#ifdef ROCKSDB_RANGESYNC_PRESENT
+
+#if !defined(ZFS_SUPER_MAGIC)
+// The magic number for ZFS was not exposed until recently. It should be fixed
+// forever so we can just copy the magic number here.
+#define ZFS_SUPER_MAGIC 0x2fc12fc1
+#endif
+
+bool IsSyncFileRangeSupported(int __attribute__((__unused__)) fd) {
+  // `fstatfs` is only available on Linux, but so is `sync_file_range`, so
+  // `defined(ROCKSDB_RANGESYNC_PRESENT)` should imply `defined(OS_LINUX)`.
+  struct statfs buf;
+  int ret = fstatfs(fd, &buf);
+  assert(ret == 0);
+  if (ret != 0) {
+    // We don't know whether the filesystem properly supports `sync_file_range`.
+    // Even if it doesn't, we don't know of any safety issue with trying to call
+    // it anyways. So, to preserve the same behavior as before this `fstatfs`
+    // check was introduced, we assume `sync_file_range` is usable.
+    return true;
+  }
+  if (buf.f_type == ZFS_SUPER_MAGIC) {
```

Comment:
I wonder if it would be better to invert this and have a whitelist of filesystems that we know support `sync_file_range` instead of a blacklist of filesystems which don't support it. Accidentally having `bytes_per_sync` not have any effect seems like the poorer failure mode.

Reply:
I tried on btrfs to see whether there is a general issue with copy-on-write filesystems (as the man-page talks about but doesn't exactly make clear, IMO), or if we're seeing a ZFS oddity. It worked on btrfs, which leads me to believe ZFS is an outlier here. Maybe it has a bug.

Reply:
Interesting. I perused the kernel sources earlier today and the `sync_file_range` implementation appears to be generic (i.e. applies to all file systems). I know that ZFS tries to replace much of the linux buffer cache code with its own code, though.